app/src/app/api/v0/token/route.ts

```diff
@@ -0,0 +1,183 @@
+import { createHash } from "node:crypto";
+import { hasFeatureFlag, FeatureFlags } from "@/util/feature-flags";
+
+const ACCESS_TOKEN_EXPIRY = 7 * 24 * 60 * 60;
+const REFRESH_TOKEN_EXPIRY = 30 * 24 * 60 * 60;
+
+function verifyPKCE(verifier: string, challenge: string): boolean {
+  const hash = createHash("sha256").update(verifier, "ascii").digest();
+  const base64url = hash
+    .toString("base64")
+    .replace(/\+/g, "-")
+    .replace(/\//g, "_")
+    .replace(/=+$/, "");
+  return base64url === challenge;
+}
+
+/**  */
+export const POST = apiHandler(async (_req, { params, session }) => {
+
+  if (!hasFeatureFlag(FeatureFlags.OAUTH_ENABLED)) {
+    throw createHttpError.InternalServerError("OAuth 2.0 not enabled");
+  }
+
+  if (!process.env.VERIFICATION_TOKEN_SECRET) {
+    logger.error("Need to assign VERIFICATION_TOKEN_SECRET in env!");
+    throw createHttpError.InternalServerError("Configuration error");
+  }
+
+  const contentType = _req.headers.get('content-type') || ''
+
+  if (!contentType.includes('application/x-www-form-urlencoded')) {
+    return NextResponse.json(
+      { error: 'unsupported_content_type' },
+      { status: 415 }
+    )
+  }
+
+  const formData = await _req.formData()
+
+  const grantType = formData.get("grant_type");
+  const code = formData.get("code");
+  const redirectUri = formData.get("redirect_uri");
+  const clientId = formData.get("client_id");
+  const codeVerifier = formData.get("code_verifier");
+
+  if (grantType == null) {
+    throw new createHttpError.BadRequest("grant_type is required")
+  }
+
+  if (typeof grantType !== 'string') {
+    throw new createHttpError.BadRequest("grant_type must be a string")
+  }
+
+  if (grantType !== "authorization_code") {
+    throw new createHttpError.BadRequest("grant_type must 'authorization_code'")
+  }
+
+  if (code == null) {
+    throw new createHttpError.BadRequest("code is required")
+  }
+
+  if (typeof code !== 'string') {
+    throw new createHttpError.BadRequest("code must be a string")
+  }
+
+  if (clientId == null) {
+    throw new createHttpError.BadRequest("client_id is required")
+  }
+
+  if (typeof clientId !== 'string') {
+    throw new createHttpError.BadRequest("client_id must be a string")
+  }
+
+  if (redirectUri == null) {
+    throw new createHttpError.BadRequest("redirect_uri is required")
+  }
+
+  if (typeof redirectUri !== 'string') {
+    throw new createHttpError.BadRequest("redirect_uri must be a string")
+  }
+
+  if (codeVerifier == null) {
+    throw new createHttpError.BadRequest("code_verifier is required")
+  }
+
+  if (typeof codeVerifier !== 'string') {
+    throw new createHttpError.BadRequest("code_verifier must be a string")
+  }
+
+  const client = await getClient(clientId);
+
+  if (!client) {
+    throw new createHttpError.BadRequest("Unrecognized client_id")
+  }
+
+  if (client.redirectUri !== redirectUri) {
+    throw new createHttpError.BadRequest("redirect_uri mismatch")
+  }
```

Comment:
Use `zod` where possible to validate instead of doing a bunch of manual null and type checks.
See e.g. `src/util/validation.ts` and:
https://zod.dev/
The advantage is that our validation errors (and their schema) are standardized across the application.

Reply:
OK, I created a task for this.